### PR TITLE
Replace modulepaths based on paths and package location

### PR DESCRIPTION
**Pull request by mokkabonna**
Work in Progress. Pretty basic for now, but works. Feedback is welcome!

As mentioned in https://github.com/agrc/atom-amdbutler/pull/17 i have done some work on this.

The way it works now is:
1. New option **requirejsConfigFile**
2. Tries to read and parse as JSON baseFolderPath + requirejsConfigFile, if not ok, fail silently. Will work as before.
3. If ok creates a replace map based on the paths and packages
4. When crawling for modules, returns the rewritten path instead of the full disk path. The path that matches first is used.

The map looks like this:

``` js
{
  "bower_components/lodash-amd/modern": "lodash",
  "bower_components/jquery/dist/jquery": "jquery"
}
```

And is sorted with the longest names first so that 

``` js
{
  "bower_components/durandal/js/plugins": "plugins"
  "bower_components/durandal/js":  "durandal"
}
```

results in "plugins/router" instead of "durandal/plugins/router"

In my app I have among others these packages:

``` js
{
  "packages": [{
    "name": "lodash",
    "location": "bower_components/lodash-amd/modern"
  }, {
    "name": "mout",
    "location": "bower_components/mout/src"
  }]
}
```

Resulting in this as an example:

image

And then imported like this:

``` js
define([
  './startApp',
  'lodash/array/flatten',
], function(
  startApp,
  flatten
) {
  'use strict';
});
```

Possible future options
- [ ] Support several requirejs paths, and will try to open them in order, similar to basepaths
- [ ] Support non JSON config, like requirejs does with it's build (beyond me why, just complicates things) Probably need to use a parser like esprima to do this correctly. For now JSON works for my usecase at least.
- [ ] Support package main files so that **lodash/main.js** file is presented only as **lodash**. Need to support the main property for differently named package mains.

### CODE REVIEW

File: lib/crawler.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ module.exports = {
 
         return path.join(filePath.split(baseFolderName)[0], baseFolderName);
     },
-    crawl: function (basePath, currentPackage) {
+    crawl: function (basePath, currentPackage, replaceMap) {
```
Comment (mokkabonna) [inline lib/crawler.js:31]: A lot of passing around replaceMap after your last changes in the crawler.

File: lib/mods-view.js [hunk 2]
```diff
@@ -62,7 +62,9 @@ ModsView.prototype.show = function (items, action, excludePaths) {
     }
     this.panel.show();
     this.setItems(items.filter(function (i) {
-        return excludePaths.indexOf(i.path) === -1;
+        return !excludePaths.some(function(pathToExclude) {
```
Comment (mokkabonna) [inline lib/mods-view.js:65]: Again just for testing now, but the idea is that if the path contains any of the strings in the exclude list, exclude it.

**Comment by stdavis** [comment]
This is looking nice. We typically define our packages from with a module [like this](https://github.com/agrc/wri-web/blob/master/src/app/run.js). But I could see us moving it to a .json file to resolve [another issue](https://github.com/agrc/AGRCJavaScriptProjectBoilerPlate/issues/118).

**Comment by mokkabonna** [comment]
We also had it in a plain js file until we started to need it in non browser environments, like our build process, and automatic inclusion of bower packages in the paths etc. Then JSON is so much simpler to deal with. Sure you could parse the js with esprima or similar, but it surely complicates things. Keeping it in JSON keeps it accessible in the app as well as in build process or like here for editor integration.

I don't see it as bad to support plain js, but is quite a different task, quite a bit more complicated. 

---

A problem with AMD is the "whatever works" approach. Meaning there is no package.json or no way to identify an AMD based project. There are so many ways to use AMD, and no standard location of config file or format of it. There exists so many ways to accomplish one thing imo.

I wish someone with influence in the community would create a AMD subset with some rules around this. Would make it easier to write tooling that used these conventions to do things like we try to do here, locate a project specific config file.

For now, maybe instead of or in addition to supporting the requirejsConfigFile setting, allow people to put a `"amdConfig":  "requirejsConfig.json"` in their bower.json? In addition to using the moduleType : ['amd'] https://github.com/bower/spec/blob/master/json.md#moduletype

For bower packages at least..

**Comment by mokkabonna** [comment]
I am working on a amd-hyperclick provider (see https://github.com/AsaAyers/js-hyperclick/issues/12), and have the same issues there, how to locate the config. Some other packages that have the same issue have decided to create their own config file, like tern js: http://ternjs.net/doc/manual.html

But I feel without a AMD only solution to this we have this situation: https://xkcd.com/927/

**Pull request closed without merging**